### PR TITLE
feat(core): Handle wallet collision while creation

### PR DESCRIPTION
**Pull request by amanCypherock**
This PR is part-1 of wallet verification through nonce.

Changes: 
- On wallet creation, a random nonce(common to each share) of 12 bytes is generated and stored on cards and the device.
- On wallet sync flow, the random nonce received from the wallet on the card is stored on flash.

### CODE REVIEW

File: src/wallet/create_new_wallet_flow.c [hunk 1]
```diff
@@ -326,8 +326,18 @@ new_wallet_state_e new_wallet_state_handler(new_wallet_state_e current_state) {
                         wallet.total_number_of_shares,
                         wallet.minimum_number_of_shares,
                         wallet_shamir_data.mnemonic_shares);
-      if (WALLET_IS_PIN_SET(wallet.wallet_info))
+
+      uint8_t wallet_nonce[NONCE_SIZE] = {0};
+      random_generate(wallet_nonce, 12);
```
Comment (ujjwal-cyph) [inline src/wallet/create_new_wallet_flow.c:331]: define a macro for a nonce size of 12

suggestion: could be NONCE_SIZE = 12 and PADDED_NONCE_SIZE = 16

**Review by ujjwal-cyph — CHANGES_REQUESTED** [review]

File: src/wallet/create_new_wallet_flow.c [hunk 2]
```diff
@@ -326,8 +326,18 @@ new_wallet_state_e new_wallet_state_handler(new_wallet_state_e current_state) {
...
+
+      for (int i = 0; i < TOTAL_NUMBER_OF_SHARES; i++) {
+        memcpy(wallet_shamir_data.share_encryption_data[i], wallet_nonce, 12);
+        wallet_shamir_data.share_encryption_data[i][15] = 0x01;
```
[the remark this answers is not in the document]
Comment (amanCypherock) [reply src/wallet/create_new_wallet_flow.c]: Abstracted to a separate function added there.

File: src/wallet/create_new_wallet_flow.c (hunk 1 above)
Comment (amanCypherock) [reply src/wallet/create_new_wallet_flow.c]: updated

File: common/libraries/util/wallet_utilities.c [hunk 3]
```diff
@@ -177,4 +177,19 @@ Card_Data_errors_t validate_wallet(Wallet *wallet) {
...
+
+void derive_wallet_nonce(
+    uint8_t share_encryption_data[TOTAL_NUMBER_OF_SHARES]
+                                 [PADDED_NONCE_SIZE + WALLET_MAC_SIZE]) {
+  uint8_t wallet_nonce[NONCE_SIZE] = {0};
+  random_generate(wallet_nonce, 12);
+
+  for (int i = 0; i < TOTAL_NUMBER_OF_SHARES; i++) {
+    // First 12 bytes of share_encryption_data are wallet nonce.
+    memcpy(share_encryption_data[i], wallet_nonce, 12);
```
Comment (ujjwal-cyph) [inline common/libraries/util/wallet_utilities.c:190]: use the MACRO `NONCE_SIZE`

**Review by ujjwal-cyph — APPROVED** [review]
LGTM

**Pull request merged**